xtask/src/codegen.rs

```diff
@@ -0,0 +1,79 @@
+use clap::Args;
+use eyre::Context;
+use std::fmt::Write;
+use std::io;
+
+use crate::util::{get_cargo_workspace, get_matrix};
+
+#[derive(Args, Debug)]
+pub struct Codegen {
+    /// Provide verbose diagnostic output.
+    #[clap(short, long)]
+    verbose: bool,
+}
+
+pub fn codegen(Codegen { .. }: Codegen) -> cross::Result<()> {
+    let path = get_cargo_workspace().join("src/docker/provided_images.rs");
+    std::fs::write(path, docker_images()).wrap_err("when writing src/docker/provided_images.rs")?;
+    Ok(())
+}
+
+pub fn docker_images() -> String {
+    let mut images = String::from(
+        r##"#![doc = "*** AUTO-GENERATED, do not touch. Run `cargo xtask codegen` to update ***"]
+use super::{ImageArchitecture, ProvidedImage};
+
+#[rustfmt::skip]
+pub static PROVIDED_IMAGES: &[ProvidedImage] = &["##,
+    );
+    let mut dir = crate::util::get_cargo_workspace().to_owned();
+    dir.push("docker");
+
+    let dir = dir.read_dir().unwrap();
+    let mut paths = dir.collect::<io::Result<Vec<_>>>().unwrap();
+    paths.sort_by_key(|e| e.path());
+    for image_target in get_matrix() {
+        write!(
+            &mut images,
+            r#"
+        ProvidedImage {{
+            name: "{}",
+            platforms: &[{}],
+            sub: {}
+        }},"#,
+            image_target.target.clone(),
+            if let Some(platforms) = &image_target.platforms {
+                platforms
+                    .iter()
+                    .map(|p| {
+                        format!(
+                            "ImageArchitecture::{}",
+                            p.replace('-', "_").to_ascii_uppercase()
+                        )
+                    })
+                    .collect::<Vec<_>>()
+                    .as_slice()
+                    .join(", ")
+            } else {
+                "ImageArchitecture::X86_64_UNKNOWN_LINUX_GNU".to_string()
```

Comment:
We re-use this a lot: can we provide a `Default` implementation instead or const value like `DEFAULT_IMAGE_ARCHITECTURE`? Since this needs to be used globally, maybe the latter is preferable? I don't see our default architecture changing, but it's not very DRY right now.